cmd/pj-rehearse-metrics/pj-rehearse-metrics.go

```diff
@@ -0,0 +1,172 @@
+	metricsPathInRun = "artifacts/rehearse-metrics.json"
+)
+
+type scraper struct {
+	cacheDir string
+	bucket   *storage.BucketHandle
+	ctx      context.Context
+}
+
+func (s *scraper) scrapeRun(runPrefix string, prNumber string) {
+	runNumber := path.Base(runPrefix)
+	metricsPath := path.Join(runPrefix, metricsPathInRun)
+
+	prDir := filepath.Join(s.cacheDir, prNumber)
+	localMetricsPath := filepath.Join(prDir, runNumber)
+	if _, err := os.Stat(localMetricsPath); err == nil {
+		return
+	}
+
+	reader, err := s.bucket.Object(metricsPath).NewReader(s.ctx)
+	if err != nil {
+		return
+	}
+	data, err := ioutil.ReadAll(reader)
+	_ = reader.Close()
+	if err != nil {
+		fmt.Printf("could not read rehearse metrics err=%v\n", err)
+		return
+	}
+
+	if err := os.MkdirAll(prDir, 0755); err != nil {
+		fmt.Printf("could not create a directory for PR err=%v\n", err)
+		return
+	}
+
+	if err := ioutil.WriteFile(localMetricsPath, data, 0644); err != nil {
+		fmt.Printf("could not write metrics to %s err=%v\n", localMetricsPath, err)
+		return
+	}
+}
+
+func (s *scraper) scrapePr(prefix string) {
+	prNumber := path.Base(prefix)
+	jobPrefix := path.Join(prefix, rehearseJobName)
+	jobPrefix = fmt.Sprintf("%s/", jobPrefix) // Needed because path.Join() removes trailing slash
+
+	runs := s.bucket.Objects(s.ctx, &storage.Query{Prefix: jobPrefix, Delimiter: "/"})
+	for {
+		attrs, err := runs.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			fmt.Printf("Next() failed err=%v\n", err)
+			break
+		}
+
+		// Leaf (a file), not a prefix (a dir)
+		if attrs.Name != "" {
+			continue
+		}
+
+		s.scrapeRun(attrs.Prefix, prNumber)
+	}
+}
+
+type options struct {
+	cacheDir string
+}
+
+func gatherOptions() options {
+	o := options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+
+	fs.StringVar(&o.cacheDir, "cache-dir", "", "Path to a directory where scraped metrics data will be saved. If not provided, cache will not be used.")
+
+	_ = fs.Parse(os.Args[1:])
+	return o
+}
+
+func main() {
+	o := gatherOptions()
+	var err error
+
+	if o.cacheDir == "" {
+		o.cacheDir, err = ioutil.TempDir("", "")
+		if err != nil {
+			_, _ = fmt.Fprintf(os.Stderr, "Failed to create a temporary directory: %v", err)
+			os.Exit(1)
+		}
+		defer os.RemoveAll(o.cacheDir)
+	}
+
+	ctx := context.Background()
+
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		_, _ = fmt.Fprintf(os.Stderr, "Failed to create client: %v", err)
+		os.Exit(1)
+	}
+
+	bucket := client.Bucket(originCiBucket)
+	pullRequests := bucket.Objects(ctx, &storage.Query{
+		Prefix:    "pr-logs/pull/openshift_release/",
+		Delimiter: "/",
+	})
+
+	scraper := scraper{ctx: ctx, bucket: bucket, cacheDir: o.cacheDir}
+	prPrefixes := make(chan string, 100)
+	progress := make(chan string, 100)
+
+	wg := sync.WaitGroup{}
+	wg.Add(20)
+
+	for w := 0; w < 20; w++ {
+		go func() {
+			for prefix := range prPrefixes {
+				scraper.scrapePr(prefix)
+				progress <- prefix
```

Comment:
this should be like:
```go
defer wg.Done()
for {
    select {
        prefix: = < -prPrefixes
        scraper.scrapePr(prefix)
        progress < -prefix
        case <-done:
            close(prPrefixes)
            close(progress)
            return
    }
}
}
```

Reply:
you need also the other channel:     done := make(chan struct{})

Reply:
Why _should_ this be like that? What's wrong? To me it looks more cluttered and needs an additional channel just to communicate the same thing that's communicated here by simply closing the channel.